executor/src/witgen/mod.rs

```diff
@@ -123,9 +123,174 @@ fn rows_are_repeating(values: &[(&str, Vec<FieldElement>)]) -> Option<usize> {
+    NoMachineMatch,
+    IdentityUnsupported,
+    Multiple(Vec<CompleteCause>),
+    None,
+}
+
+impl CompleteCause {
+    fn combine(self, right: CompleteCause) -> CompleteCause {
+        match (self, right) {
+            (CompleteCause::Multiple(l), CompleteCause::Multiple(r)) => {
+                CompleteCause::Multiple(l.into_iter().chain(r).collect())
+            }
+            (m @ CompleteCause::Multiple(_), other) | (other, m @ CompleteCause::Multiple(_)) => {
+                m.combine(CompleteCause::Multiple(vec![other]))
+            }
+            (l, r) => CompleteCause::Multiple(vec![l, r]),
+        }
+    }
+}
+
+#[derive(Clone, Debug, PartialEq, Eq)]
+pub enum IncompleteCause {
+    /// Previous value of witness column not known when trying to derive a value in the next row.
+    PreviousValueUnknown(String),
+    Generic(String),
+    Multiple(Vec<IncompleteCause>),
+}
+
+impl IncompleteCause {
+    fn combine(self, right: IncompleteCause) -> IncompleteCause {
+        match (self, right) {
+            (IncompleteCause::Multiple(l), IncompleteCause::Multiple(r)) => {
+                IncompleteCause::Multiple(l.into_iter().chain(r).collect())
+            }
+            (m @ IncompleteCause::Multiple(_), other)
+            | (other, m @ IncompleteCause::Multiple(_)) => {
+                m.combine(IncompleteCause::Multiple(vec![other]))
+            }
+            (l, r) => IncompleteCause::Multiple(vec![l, r]),
+        }
+    }
+}
+
+impl From<String> for IncompleteCause {
+    fn from(value: String) -> Self {
+        IncompleteCause::Generic(value)
+    }
+}
+
+impl Display for IncompleteCause {
+    fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
+        use IncompleteCause::*;
+        match self {
+            Generic(reason) => write!(f, "{reason}"),
+            PreviousValueUnknown(names) => write!(
+                f,
+                "Previous value of the following column(s) is not (yet) known: {names}.",
+            ),
+            Multiple(errors) => {
+                let (previous_unknown, mut others) = errors.iter().fold(
+                    (vec![], vec![]),
+                    |(mut previous_unknown, mut others), err| {
+                        if let PreviousValueUnknown(n) = err {
+                            previous_unknown.push(n.clone());
+                        } else {
+                            others.push(format!("{err}"));
+                        }
+                        (previous_unknown, others)
+                    },
+                );
+                if !previous_unknown.is_empty() {
+                    others.push(format!(
+                        "{}",
+                        PreviousValueUnknown(previous_unknown.join(", "))
+                    ));
+                }
+                write!(f, "{}", others.join("\n"))
+            }
+        }
+    }
+}
+
+type Constraints = Vec<(usize, Constraint)>;
+
+#[derive(Clone, Debug, PartialEq, Eq)]
+enum Complete {
```

Comment:
You sure about `Complete::Incomplete`?

Reply:
I'm thinking of changing to `Status`

Reply:
Let's see after I remove `CompleteCause` entirely